pkg/visitors/assemble.go

```diff
@@ -71,6 +72,86 @@ func (v *Assemble) Visit(f uefi.Firmware) error {
 		}
 
 		fileOffset := f.DataOffset
+
+		var pad *uefi.File
+		var shrink uint64
+		for _, file := range f.Files {
+			fileBuf := file.Buf()
+			fileLen := uint64(len(fileBuf))
+			if fileLen == 0 {
+				log.Fatal("Unexpected empty file buffer (at least a header was expected): %v", file.Header.GUID)
+			}
+			if file.Header.GUID == *uefi.FFGUID && fileLen > 24 {
+				// found non-empty PAD to check for free space.
+				pad = file
+			}
+
+			// Pad to the 8 byte alignments.
+			alignedOffset := uefi.Align8(fileOffset)
+			// Read out the file alignment requirements
+			if alignBase := file.Header.Attributes.GetAlignment(); alignBase != 1 {
+				hl := file.HeaderLen()
+				// We need to align the data, not the header. This is so terrible.
+				fileDataOffset := uefi.Align(alignedOffset+hl, alignBase)
+				// Calculate the starting offset of the file
+				newOffset := fileDataOffset - hl
+				if gap := (newOffset - alignedOffset); gap >= 8 && gap < uefi.FileHeaderMinLength {
+					// We need to re align to the next boundary cause we can't put a pad file in here.
+					// Who thought this was a good idea?
+					fileDataOffset = uefi.Align(fileDataOffset+1, alignBase)
+					newOffset = fileDataOffset - hl
+				}
+				if newOffset != alignedOffset {
+					// Add a pad file starting from alignedOffset to newOffset
+					alignedOffset = newOffset
+				}
+			}
+			fileOffset = alignedOffset + fileLen
+		}
+
+		// Check if we're out of space.
+		if f.Length < fileOffset && !f.Resizable {
+			// do we have a non-empty pad file?
+			if pad != nil {
+				shrink = fileOffset - f.Length
+				trim := shrink
+				if trim > (uint64(len(pad.Buf()) + 24)) {
+					return fmt.Errorf("out of space in non-empty padding. space available: %v bytes, need apace: %v bytes", len(pad.Buf()), trim)
+				}
+				// remove free space after the header from pad file and recalculate
+				tmp_header := pad.Buf()[:24]
+				tmp_body := pad.Buf()[trim:]
+				tmp_state := tmp_header[23]
+				tmp_size := len(tmp_header) + len(tmp_body)
+				tmp_header[20] = uint8((tmp_size) & 0xff)
+				tmp_header[21] = uint8((tmp_size / 0x100) & 0xff)
+				tmp_header[22] = uint8((tmp_size / 0x10000) & 0xff)
+				tmp_header[16] = 0x0
+				tmp_header[17] = 0x0
+				tmp_header[23] = 0x0
+				tmp_header[16] = uint8((uint16(0x100) - uint16(uefi.Checksum8(tmp_header))) & 0xff)
+				tmp_header[23] = tmp_state
+				tmp_header[17] = 0xaa // Fixed
+				pad.SetBuf(append(tmp_header, tmp_body...))
+			} else {
+				fmt.Printf("no non-empry pad file\n")
+			}
+		}
+
+		// Second pass to check to see if there is enough free space in FV or non-empty pad file
```

Comment:
I'm hesitant to try to find space in non empty pad files in the middle of the volume, because those pad files may be placed there to satisfy alignment requirements, and moving them around can cause things to silently break.

Reply:
I'm not moving the pad file around per se, but shrink it to fit in an EFI driver at the original starting address of the pad file.  This way the ending point is the same.  It is critical that we don't change the ending offset or in my use model the BIOS won't boot since the entry point will have a different address :-(  In the last FV, it is laid out as:
   PEI module
   PEC core
   PEI module
   ....
   Freefrom
   ....
   PEI module
   ...
   Pad
   Raw
   ...
   Pad (non-empty) 
   SEC core
then after the insert
   PEI module
   PEC core
   PEI module
   ....
   Freefrom
   ....
   PEI module
   ...
   Pad
   Raw
   ...
   DXE driver -- new EFI driver inserted
   Pad (non-empty) size is now original size - inserted DXE driver
   SEC core
So the ending address of the pad file hasn't changed and the starting address of the inserted DXE is the same as the original pad file.  The starting address of the pad file is now at the end of the inserted DXE driver.

In one of my use cases the original pad file was 1096544 and is now 853640 with the inserted file 242902 according UEFITool Full size values.  If the ending address of the pad file changes then the system won't boot.

Reply:
I see, I think I understand your problem a bit better. Is it correct that PEI, DXE and even SEC are all in the same FV in your image? Hmm let me try and cook up something over the weekend to see if I can do this differently

Reply:
DXE could be in the this FV, typically they are in the DXE volume but when the free space is used up in the DXE FV then they can spill over into the PEI FV.  That FV doesn't have free space by itself but appears to wrap the free space inside the pad file.  I put some utk table dumps at:
     http://www.ambrisko.com/doug/table.orig
     http://www.ambrisko.com/doug/table.customized
The customized one had 3 DXE drivers added with GUID of type
     9F3A0016-AE55-4288-829D-D22FD344C34*
The last one was inserted into the pad file area
@@ -1591,7 +1596,9 @@ Image                                                 
    File           FFFFFFFF-FFFF-FFFF-FFFF-FFFFFFFFFFFF  EFI_FV_FILETYPE_FFS_PAD                0x18
    File           17088572-377F-44EF-8F4E-B09FFF46A070  EFI_FV_FILETYPE_RAW                 0x27028
    File           C3E36D09-8294-4B97-A857-D5288FE33E28  EFI_FV_FILETYPE_RAW                    0x52
-   File           FFFFFFFF-FFFF-FFFF-FFFF-FFFFFFFFFFFF  EFI_FV_FILETYPE_FFS_PAD            0x10bb60
+   File           9F3A0016-AE55-4288-829D-D22FD344C344  EFI_FV_FILETYPE_DRIVER              0x3b4d6
+    Sec                                                 EFI_SECTION_COMPRESSION             0x3b4be
+   File           FFFFFFFF-FFFF-FFFF-FFFF-FFFFFFFFFFFF  EFI_FV_FILETYPE_FFS_PAD             0xd0688
    File           1BA0062E-C779-4582-8566-336AE8F78F09  EFI_FV_FILETYPE_SECURITY_CORE        0x2148
     Sec                                                 EFI_SECTION_RAW                         0x4
     Sec                                                 EFI_SECTION_PE32                     0x20e4